### PR TITLE
[istio] add a MutatingWebhook to inject istio-proxy to individual pods

### PR DESCRIPTION
**Pull request by fl64**
## Description
Istio global mutating webhook allows injecting istio sidecars into individual pods. 
Redesigned logic for enabling istio-sidecar for ingress nginx module.

## Why do we need it, and what problem does it solve?
Closes https://github.com/deckhouse/deckhouse/issues/2257

## What is the expected result?
Set the `sidecar.istio.io/inject: true` label to inject istio sidecar to the pod.
 
## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: istio
type: chore
summary: |
- Istio global mutating webhook allows injecting istio sidecars into individual Pods. 
- Redesigned logic for enabling istio-sidecar for the `ingress-nginx` module.
impact_level: default
```

### CODE REVIEW

File: ee/modules/110-istio/docs/README.md [hunk 1]
```diff
@@ -184,6 +184,8 @@ The sidecar-injector is a recommended way to add sidecars. Istio can inject side
   * `istio.io/rev=v1x13` — use the specific Istio version for a given namespace.
 * The `sidecar.istio.io/inject` (`"true"` or `"false"`) **Pod** annotation lets you redefine the `sidecarInjectorPolicy` policy locally. These annotations work only in namespaces to which the above labels are attached.
 
+It is also possible to add sidecar to an individual pod in namespace without the `istio-injection=renabled` or `istio.io/rev=vXxYZ` labels, by setting the `sidecar.istio.io/inject: true` label.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/README.md:187]:
```suggestion
It is also possible to add sidecar to an individual pod in namespace without the `istio-injection=enabled` or `istio.io/rev=vXxYZ` labels, by setting the `sidecar.istio.io/inject: true` label.
```

File: modules/402-ingress-nginx/templates/controller/controller.yaml [hunk 2]
```diff
@@ -133,16 +133,18 @@ spec:
   template:
     metadata:
       labels:
+      {{- if include "is_istio_in_use" $context }}
```
Comment (apolovov) [inline modules/402-ingress-nginx/templates/controller/controller.yaml:136]: get rid of is_istio_in_use pls, it was a mistake.

File: ee/modules/110-istio/docs/README_RU.md [hunk 3]
```diff
@@ -186,6 +186,8 @@ Istio позволяет осуществлять сбор трейсов с п
   * `istio.io/rev=v1x13` — использовать конкретную версию Istio для данного namespace.
 * Аннотация к **Pod'у** — `sidecar.istio.io/inject` (`"true"` или `"false"`), позволяет локально переопределить политику `sidecarInjectorPolicy`. Эти аннотации работают только в namespace, обозначенных лейблами из списка выше.
 
+Также существует возможность добавить sidecar к индивидуальному поду в namespace без установленных лейблов `istio-injection=enabled` или `istio.io/rev=vXxYZ`, путем установки лейбла `sidecar.istio.io/inject=true`
```
Comment (apolovov) [inline ee/modules/110-istio/docs/README_RU.md:189]:
```suggestion
Также существует возможность добавить sidecar к индивидуальному поду в namespace без установленных лейблов `istio-injection=enabled` или `istio.io/rev=vXxYZ` путём установки лейбла `sidecar.istio.io/inject=true`.
```

File: ee/modules/110-istio/docs/README.md [hunk 4]
```diff
@@ -184,6 +184,8 @@ The sidecar-injector is a recommended way to add sidecars. Istio can inject side
   * `istio.io/rev=v1x13` — use the specific Istio version for a given namespace.
 * The `sidecar.istio.io/inject` (`"true"` or `"false"`) **Pod** annotation lets you redefine the `sidecarInjectorPolicy` policy locally. These annotations work only in namespaces to which the above labels are attached.
 
+It is also possible to add sidecar to an individual pod in namespace without the `istio-injection=enabled` or `istio.io/rev=vXxYZ` labels, by setting the `sidecar.istio.io/inject=true` label.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/README.md:187]:
```suggestion
It is also possible to add the sidecar to an individual pod in namespace without the `istio-injection=enabled` or `istio.io/rev=vXxYZ` labels by setting the `sidecar.istio.io/inject=true` label.
```

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml [hunk 5]
```diff
@@ -1,23 +1,39 @@
...
-    service:
-      name: istiod-{{ .Values.istio.internal.globalRevision }}
-      namespace: d8-istio
-      path: /inject
-      port: 443
-  failurePolicy: Fail
-  matchPolicy: Exact
-  name: sidecar-injector.istio.io
+{{- /* Case 1: Namespace selector enabled, and object selector is not injected */}}
+{{- include "core" (mergeOverwrite (deepCopy $whv) (dict "Prefix" "namespace-") ) }}
```
Comment (apolovov) [inline ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml:36]: 😮

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml (hunk 5 above)
Comment (apolovov) [reply ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml:36]: {{- include "mwh_common" (list . "namespace") -}}

File: ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml [hunk 6]
```diff
@@ -32,16 +48,20 @@ webhooks:
...
-    apiVersions:
-    - v1
-    operations:
-    - CREATE
-    resources:
-    - pods
-    scope: '*'
-  sideEffects: None
-  timeoutSeconds: 30
+{{- /* Case 2: no namespace label, but object selector is enabled (and revision label is not, which has priority) */}}
```
Comment (apolovov) [inline ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml:51]: To think.

**Review by apolovov — CHANGES_REQUESTED** [review]

File: ee/modules/110-istio/docs/README.md [hunk 7]
```diff
@@ -184,6 +184,8 @@ The sidecar-injector is a recommended way to add sidecars. Istio can inject side
   * `istio.io/rev=v1x13` — use the specific Istio version for a given namespace.
 * The `sidecar.istio.io/inject` (`"true"` or `"false"`) **Pod** annotation lets you redefine the `sidecarInjectorPolicy` policy locally. These annotations work only in namespaces to which the above labels are attached.
 
+It is also possible to add the sidecar to an individual pod in namespace without the `istio-injection=enabled` or `istio.io/rev=vXxYZ` labels by setting the `sidecar.istio.io/inject=true` label.
```
Comment (apolovov) [inline ee/modules/110-istio/docs/README.md:187]:
```suggestion
It is also possible to add the sidecar to an individual pod in namespace without the `istio-injection=enabled` or `istio.io/rev=vXxYZ` labels by setting the `sidecar.istio.io/inject=true` Pod label.
```

File: ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml [hunk 8]
```diff
@@ -1,23 +1,37 @@
+{{- define "mwh_common" }}
+  {{- $context := index . 0 }} {{- /* Dot object (.) with .Values, .Chart, etc */ -}}
+  {{- $prefix := index . 1 }}  {{- /* argv1 */ -}}
```
Comment (apolovov) [inline ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml:3]:
```suggestion
  {{- $context := index . 0 }}
  {{- $prefix := index . 1 }}
```

File: ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml [hunk 9]
```diff
@@ -1,23 +1,37 @@
...
-    caBundle: {{ .Values.istio.internal.ca.cert | b64enc }}
-    service:
-      name: istiod-{{ .Values.istio.internal.globalRevision }}
-      namespace: d8-istio
-      path: /inject
-      port: 443
-  failurePolicy: Fail
-  matchPolicy: Exact
-  name: sidecar-injector.istio.io
+{{- /* Case 1: Injection for namespace */}}
```
Comment (apolovov) [inline ee/modules/110-istio/templates/control-plane/mutatingwebhook-global.yaml:33]:
```suggestion
{{- /* Case 1: Namespace-wide injection */}}
```

File: ee/modules/110-istio/monitoring/prometheus-rules/revisions.yaml [hunk 10]
```diff
@@ -81,7 +81,8 @@
       plk_protocol_version: "1"
       summary: There are Pods with istio sidecars, but without istio-injection configured
     expr: |
-      d8_istio_pod_revision{revision!="absent", desired_revision="absent"}
+      d8_istio_pod_revision{desired_revision="absent",revision!="absent"}
+      unless on (dataplane_pod) label_replace(kube_pod_labels{label_sidecar_istio_io_inject="true"}, "dataplane_pod", "$1", "pod", "(.+)")
```
Comment (apolovov) [inline ee/modules/110-istio/monitoring/prometheus-rules/revisions.yaml:85]: this logic should be in revisions_monitoring hook.

Comment (fl64) [reply ee/modules/110-istio/monitoring/prometheus-rules/revisions.yaml]: fixed

File: ee/modules/110-istio/docs/README_RU.md [hunk 11]
```diff
@@ -186,6 +186,8 @@ Istio позволяет осуществлять сбор трейсов с п
   * `istio.io/rev=v1x13` — использовать конкретную версию Istio для данного namespace.
 * Аннотация к **Pod'у** — `sidecar.istio.io/inject` (`"true"` или `"false"`), позволяет локально переопределить политику `sidecarInjectorPolicy`. Эти аннотации работают только в namespace, обозначенных лейблами из списка выше.
 
+Также существует возможность добавить sidecar к индивидуальному поду в namespace без установленных лейблов `istio-injection=enabled` или `istio.io/rev=vXxYZ` путём установки лейбла `sidecar.istio.io/inject=true`.
```
Comment (z9r5) [inline ee/modules/110-istio/docs/README_RU.md:189]:
```suggestion
Также существует возможность добавить sidecar к индивидуальному Pod'у в namespace без установленных лейблов `istio-injection=enabled` или `istio.io/rev=vXxYZ` путём установки лейбла `sidecar.istio.io/inject=true`.
```

**Review by z9r5 — APPROVED** [review]

**Review by apolovov — CHANGES_REQUESTED** [review]
Detect namespaces with labeled Pods to distribute registry secret.

**Review by apolovov — APPROVED** [review]

**Review by z9r5 — APPROVED** [review]

**Pull request merged**